src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -887,6 +895,61 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
 void ClusterTaskManager::ScheduleAndDispatchTasks() {
   SchedulePendingTasks();
   DispatchScheduledTasksToWorkers(worker_pool_, leased_workers_);
+  // TODO(swang): Spill from waiting queue first? Otherwise, we may end up
+  // spilling a task whose args are already local.
+  SpillWaitingTasks();
+}
+
+void ClusterTaskManager::SpillWaitingTasks() {
+  // Try to spill waiting tasks to a remote node, prioritizing those at the end
+  // of the queue. Waiting tasks are spilled if there are enough remote
+  // resources AND (we have no resources available locally or we cannot fetch
+  // their dependencies due to lack of object store memory). We consider object
+  // store memory separately from the task's placement resources since object
+  // store memory is managed separately from other resources.
+  //
+  // NOTE(swang): We do not iterate by scheduling class here, so if we break
+  // due to lack of remote resources, it is possible that a waiting task that
+  // is earlier in the queue could have been scheduled to a remote node.
+  auto it = waiting_task_queue_.end();
+  while (it != waiting_task_queue_.begin()) {
+    it--;
+    const auto &task = std::get<0>(*it);
+    const auto &task_id = task.GetTaskSpecification().TaskId();
+    RAY_LOG(DEBUG) << "Attempting to spill back waiting task " << task_id
+                   << " to remote node";
+    // Check whether this task's dependencies are not being pulled due to lack
+    // of memory. If this is true, then we should force the task onto a remote
+    // feasible node, even if we have enough resources available locally for
+    // placement.
+    bool out_of_memory =
+        task_dependency_manager_.TaskDependenciesBlockedDueToOom(task_id);
+    auto placement_resources =
+        task.GetTaskSpecification().GetRequiredPlacementResources().GetResourceMap();
+    int64_t _unused;
+    bool is_infeasible;
+    // TODO(swang): The policy currently does not account for object store
+    // memory availability. Ideally, we should pick the node with the most
```

Comment:
Isn't it causing flooding? I think it'd be okay to just do random spilling like now